### PR TITLE
Support MOI.ScalarNonlinearFunction

### LINKED ISSUE #233: Add support for ScalarNonlinearFunction
See #206 
See #208

### PR DESCRIPTION
**Pull request by odow**
Closes #206
Closes #208 
Closes #233

Still a big WIP. I think I need to fix the LP/MIP distinction first.

### CODE REVIEW

File: src/MOI/MOI_wrapper.jl [hunk 1]
```diff
@@ -1278,6 +1287,155 @@ function MOI.modify(
...
+    f::MOI.ScalarNonlinearFunction,
+)
+    parsed, type, value = Cint(1), Cint[], Cdouble[]
+    _reverse_polish(model, f, type, value)
+    reverse!(type)
+    reverse!(value)
+    push!(type, Lib.XPRS_TOK_EOF)
+    push!(value, 0.0)
+    @show type, value
+    @checked Lib.XPRSnlpdelobjformula(model.inner)
```
Comment (odow) [inline src/MOI/MOI_wrapper.jl:1302]: @joaquimg, where did
https://github.com/jump-dev/Xpress.jl/blob/33737734c64fc08606d6ddc3581fd897cb58539c/src/lib.jl#L2289-L2291
come from? It's not in `xprs.h`

Comment (odow) [reply src/MOI/MOI_wrapper.jl:1302]: There's a symbol for it, but no documentation:
```
(base) oscar@Oscars-MBP Xpress % nm /Applications/FICO\ Xpress/xpressmp/lib/libxprs.dylib | grep XPRSnlpdelobjformula
0000000000229df0 T _XPRSnlpdelobjformula
000000000022a090 T _XPRSnlpdelobjformula_unchecked
```

File: src/MOI/MOI_wrapper.jl [hunk 2]
```diff
@@ -3582,21 +3650,28 @@ function MOI.get(
     return sort!(indices; by = x -> x.value)
 end
 
+_function_enums(::Type{<:MOI.ScalarAffineFunction}) = (AFFINE,)
+_function_enums(::Type{<:MOI.ScalarQuadraticFunction}) = (QUADRATIC,)
+_function_enums(::Type{<:MOI.ScalarNonlinearFunction}) = (SCALAR_NONLINEAR,)
+_function_enums(::Type{<:MOI.VectorAffineFunction}) = (INDICATOR,)
+_function_enums(::Type{<:MOI.VectorOfVariables}) = (SOS, RSOC)
```
Comment (odow) [inline src/MOI/MOI_wrapper.jl:3657]:
```suggestion
_function_enums(::Type{<:MOI.VectorOfVariables}) = (SOC, RSOC)
```

File: src/MOI/MOI_wrapper.jl (hunk 1 above)
Comment (joaquimg) [reply src/MOI/MOI_wrapper.jl]: I think we need the 9.0+ header for this.

Comment (joaquimg) [reply src/MOI/MOI_wrapper.jl]: it is not in 8.5, nor 8.14 not 9.3.
Maybe its something that was removed from the main API

Comment (odow) [reply src/MOI/MOI_wrapper.jl]: I have 42.01.05

Comment (odow) [reply src/MOI/MOI_wrapper.jl]: image

Web has nothing

Comment (joaquimg) [reply src/MOI/MOI_wrapper.jl]: I think I found it in 9.0.2. : https://github.com/jump-dev/Xpress.jl/commit/549073a52bd7112ebe0e50c2e5fb54580d1ea14a (see commit name)
So it was probably deprecated

Comment (odow) [reply src/MOI/MOI_wrapper.jl]: Cool cool. Testing with v8 has made this much easier to avoid breakage.

**Comment by odow** [comment]
Okie dokie. This is ready for a review.

**Comment by odow** [comment]
Okay, I've seen this failure a few times:

image

**Comment by odow** [comment]
Merging into master to prepare for a new release.

**Pull request merged**